File: rgen/src/main/OPI/R/Icare.r
# Open Perimetry Interface implementation for Icare
#
# Copyright [2022] [Andrew Turpin & Ivan Marin-Franch]
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require(rjson)

    # environment for this machine in R
if (exists(".opi_env") && !exists("Icare", where = .opi_env))
    assign("Icare", new.env(), envir = .opi_env)

#' Implementation of opiInitialise for the Icare machine.
#'
#' This is for internal use only. Use [opiInitialise()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param ipMonitor IP Address of the OPI monitor.
#' @param portMonitor TCP port of the OPI monitor.
#' @param ip IP Address of the OPI machine.
#' @param port TCP port of the OPI machine.
#'
#' @return a list contianing:
#'  * res JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - res$error Error code '0' if all good, '1' something wrong.
#'    - res$msg The success or error message.
#'
#' @examples
#' chooseOpi("Icare")
#' result <- opiInitialise(null)
#'
#' @seealso [opiInitialise()]
#'
opiInitialise_for_Icare <- function(ipMonitor = NULL, portMonitor = NULL, ip = NULL, port = NULL) {
    assign("socket", open_socket(ipMonitor, portMonitor), .opi_env$Icare)

    msg <- list(command = "choose", machine = "Icare")
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Jovp$socket)
    res <- readLines(.opi_env$Jovp$socket, n = 1)
    res <- rjson::fromJSON(res)

    msg <- list(ipMonitor = ipMonitor, portMonitor = portMonitor, ip = ip, port = port)
    msg <- c(list(command = "initialize"), msg)
    msg <- msg[!unlist(lapply(msg, is.null))]
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Icare$socket)

    res <- rjson::fromJSON(readLines(.opi_env$Icare$socket, n = 1))
    return(res)
}

#' Implementation of opiQueryDevice for the Icare machine.
#'
#' This is for internal use only. Use [opiQueryDevice()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#'
#'
#' @return a list contianing:
#'  * res JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - res$error '0' if success, '1' if error.
#'    - res$msg The error message or a structure with the following data.
#'
#' @examples
#' chooseOpi("Icare")
#' result <- opiQueryDevice(null)
#'
#' @seealso [opiQueryDevice()]
#'
opiQueryDevice_for_Icare <- function() {
if(!exists(".opi_env") || !exists("Icare", envir = .opi_env) || !("socket" %in% names(.opi_env$Icare)) || is.null(.opi_env$Icare$socket))
    stop("Cannot call opiQueryDevice without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list()
    msg <- c(list(command = "query"), msg)
    msg <- msg[!unlist(lapply(msg, is.null))]
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Icare$socket)

    res <- rjson::fromJSON(readLines(.opi_env$Icare$socket, n = 1))
    return(res)
}

#' Implementation of opiSetup for the Icare machine.
#'
#' This is for internal use only. Use [opiSetup()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param fixShape Fixation target type for eye.
#' @param fixCx x-coordinate of fixation target (degrees): Only valid values are
#'                 -20, -6, -3, 0, 3, 6, 20 for fixation type 'spot' and -3, 0, 3
#'                 for fixation type 'square'.
#' @param tracking Whether to correct stimulus location based on eye position.
#'
#' @return a list contianing:
#'  * res JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - res$error '0' if success, '1' if error.
#'    - res$msg The error message or a structure with the result of QUERY OPI
#'                 command.
#'
#' @examples
#' chooseOpi("Icare")
#' result <- opiSetup(settings = list(null))
#'
#' @seealso [opiSetup()]
#'
opiSetup_for_Icare <- function(settings = list(fixShape = NULL, fixCx = NULL, tracking = NULL)) {
if(!exists(".opi_env") || !exists("Icare", envir = .opi_env) || !("socket" %in% names(.opi_env$Icare)) || is.null(.opi_env$Icare$socket))
    stop("Cannot call opiSetup without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list(fixShape = settings$fixShape, fixCx = settings$fixCx, tracking = settings$tracking)
    msg <- c(list(command = "setup"), msg)
    msg <- msg[!unlist(lapply(msg, is.null))]
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Icare$socket)

    res <- rjson::fromJSON(readLines(.opi_env$Icare$socket, n = 1))
    return(res)
}

#' Implementation of opiPresent for the Icare machine.
#'
#' This is for internal use only. Use [opiPresent()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param x x co-ordinates of stimulus (degrees).
#' @param y y co-ordinates of stimulus (degrees).
#' @param lum Stimuli luminance (cd/m^2).
#' @param t Presentation time (ms).
#' @param w Response window (ms).
#'
#' @return a list contianing:
#'    - res$error '0' if success, '1' if error.
#'    - res$msg Error message or a structure with the following fields.
#'    - res$msg$seen '1' if seen, '0' if not.
#'    - res$msg$time Response time from stimulus onset if button pressed (ms).
#'  * res JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - res$eyex x co-ordinates of pupil at times eyet (pixels).
#'    - res$eyey y co-ordinates of pupil at times eyet (pixels).
#'    - res$eyed Diameter of pupil at times eyet (mm).
#'    - res$eyet Time of (eyex, eyey) pupil from stimulus onset (ms).
#'    - res$time_rec Time since 'epoch' when command was received at Compass or
#'                      Maia (ms).
#'    - res$time_resp Time since 'epoch' when stimulus response is received, or
#'                       response window expired (ms).
#'    - res$num_track_events Number of tracking events that occurred during
#'                              presentation.
#'    - res$num_motor_fails Number of times motor could not follow fixation
#'                             movement during presentation.
#'
#' @examples
#' chooseOpi("Icare")
#' result <- opiPresent(stim = list(null))
#'
#' @seealso [opiPresent()]
#'
opiPresent_for_Icare <- function(stim = list(x = NULL, y = NULL, lum = NULL, t = NULL, w = NULL)) {
if(!exists(".opi_env") || !exists("Icare", envir = .opi_env) || !("socket" %in% names(.opi_env$Icare)) || is.null(.opi_env$Icare$socket))
    stop("Cannot call opiPresent without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list(x = stim$x, y = stim$y, lum = stim$lum, t = stim$t, w = stim$w)
    msg <- c(list(command = "present"), msg)
    msg <- msg[!unlist(lapply(msg, is.null))]
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Icare$socket)

    res <- rjson::fromJSON(readLines(.opi_env$Icare$socket, n = 1))
    return(res)
}

#' Implementation of opiClose for the Icare machine.
#'
#' This is for internal use only. Use [opiClose()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#'
#'
#' @return a list contianing:
#'  * res JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - res$error '0' if success, '1' if error.
#'    - res$msg The error message or additional results from the CLOSE command
#'    - res$time The time stamp for fixation data
#'    - res$x The time stamp for fixation data
#'    - res$y The time stamp for fixation data
#'
#' @examples
#' chooseOpi("Icare")
#' result <- opiClose(null)
#'
#' @seealso [opiClose()]
#'
opiClose_for_Icare <- function() {
if(!exists(".opi_env") || !exists("Icare", envir = .opi_env) || !("socket" %in% names(.opi_env$Icare)) || is.null(.opi_env$Icare$socket))
    stop("Cannot call opiClose without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list()
    msg <- c(list(command = "close"), msg)
    msg <- msg[!unlist(lapply(msg, is.null))]
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Icare$socket)

    res <- rjson::fromJSON(readLines(.opi_env$Icare$socket, n = 1))
    return(res)
}


#' Set background color and luminance in both eyes.
#' Deprecated for OPI >= v3.0.0 and replaced with [opiSetup()].
#' @usage NULL
#' @seealso [opiSetup()]
opiSetBackground_for_Icare <- function(lum, color, ...) {return("Deprecated")}

File: core/src/main/java/org/lei/opi/core/Icare.java
package org.lei.opi.core;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.Base64;
import java.util.HashMap;

import org.apache.commons.lang3.ArrayUtils;
import org.lei.opi.core.definitions.Packet;
import org.lei.opi.core.definitions.Parameter;
import org.lei.opi.core.definitions.ReturnMsg;

import javafx.application.Platform;
import javafx.scene.Scene;
import javafx.scene.image.Image;

public abstract class Icare extends OpiMachine {
  
    /** Allowed eye values */
    public enum Eye {LEFT, RIGHT}
    /** Allowed fixation types */
    public enum Fixation {SPOT, SQUARE}
  
    /** {@value OPI_OPEN} */
    private static final String OPI_OPEN = "OPI-OPEN";
    /** {@value OPI_SET_FIXATION} */
    private static final String OPI_SET_FIXATION = "OPI-SET-FIXATION ";
    /** {@value OPI_SET_TRACKING} */
    private static final String OPI_SET_TRACKING = "OPI-SET-TRACKING ";
    /** {@value OPI_PRESENT_STATIC} */
    private static final String OPI_PRESENT_STATIC = "OPI-PRESENT-STATIC ";
    /** {@value OPI_CLOSE} */
    private static final String OPI_CLOSE = "OPI-CLOSE";
    /** {@value BAD_OPEN} */
    private static final String BAD_OPEN = "Bad open";
    /** {@value INVALID_FIXATION_SETTING} */
    private static final String INVALID_FIXATION_SETTING = "Fixation position %s is invalid for fixation type %s";
    /** {@value INVALID_TRACKING_SETTING} */
    private static final String INVALID_TRACKING_SETTING = "Tracking can only have value 0 (false) and 1 (true). It has ";
    /** {@value OPI_OPEN_FAILED} */
    private static final String OPI_OPEN_FAILED = "Problem with opening connection to iCare device. Probably need to restart.";
    /** {@value OPI_SET_FIXATION_FAILED} */
    private static final String OPI_SET_FIXATION_FAILED = "Problem with OPI-SET-FIXATION";
    /** {@value OPI_SET_TRACKING_FAILED} */
    private static final String OPI_SET_TRACKING_FAILED = "Problem with OPI-SET-TRACKING";
    /** {@value OPI_PRESENT_FAILED} */
    private static final String OPI_PRESENT_FAILED = "Problem with OPI-PRESENT";
  
    /** PRL x-coordinate measured on successful OPI-OPEN */
    private float prlx;
    /** PRL y-coordinate measured on successful OPI-OPEN */
    private float prly;
    /** ONH x-coordinate measured on successful OPI-OPEN */
    private float onhx;
    /** ONH y-coordinate measured on successful OPI-OPEN */
    private float onhy;
    /** IR Image captured on successful OPI-OPEN */
    private byte[] image;

    public static class Settings extends OpiMachine.Settings {
      public double minX;
      public double maxX;
      public double minY;
      public double maxY;
      public int minPressentationTime;
      public int maxPressentationTime;
      public int minResponseWindow;
      public int maxResponseWindow;
      public double backgroundLuminance;
      public double minLuminance;
      public double maxLuminance;
      public boolean tracking;
    };
    private Settings settings;
    public Settings getSettings() { return this.settings; }

    /*
     * @param parentScene The Scene to return to when this object is closed.
     *                    If null, then do not create a connection. (Used for GUI to probe class.)
     */
    public Icare(Scene parentScene) {
        super(parentScene);
        this.settings = (Settings) OpiMachine.fillSettings(this.getClass().getSimpleName());

        setVFCanvas(true, settings.tracking);

        if (parentScene != null)
          this.connect(settings.ip, settings.port);
    }
  
    /**
     * opiInitialise: initialize OPI
     * 
     * @param args A map of name:value pairs for Params
     * 
     * @return A JSON object with prl (x,y), onh (x,y) and Base64 encoded image
     * 
     * @since 3.0.0
     */
    @ReturnMsg(name = "res.prlx", desc = "The x-coordinate of the PRL measured initialisation (pixels).")
    @ReturnMsg(name = "res.prly", desc = "The y-coordinate of the PRL measured initialisation (pixels).")
    @ReturnMsg(name = "res.onhx", desc = "The x-coordinate of the ONH measured initialisation (pixels).")
    @ReturnMsg(name = "res.onhy", desc = "The y-coordinate of the ONH measured initialisation (pixels).")
    @ReturnMsg(name = "res.image", desc = "The captured IR image at initialisation (base64 encoded).")
    public Packet initialize(HashMap<String, Object> args) {
        try {
            System.out.println("outgoing " + this.outgoing);
            this.send(OPI_OPEN);

                // Need to temporarily read this.incoming with a DataInputStream (seems very risky!)
                // Is it possible for some other thread to read this.incoming while we are in this code chunk?
            int n = this.incoming.readInt();
            if (n < 4 * 4) // 4 floats for prlx, y, onhx, y 
                return Packet.error(OPI_OPEN_FAILED + " Did not get big enough 'n' back." );

            this.prlx = this.incoming.readFloat();
            this.prly = this.incoming.readFloat();
            this.onhx = this.incoming.readFloat();
            this.onhy = this.incoming.readFloat();

            this.image = null;
            int imageSize = n - 4 * 4;
            if (imageSize > 0) {
                this.image = new byte[imageSize];
                this.incoming.read(this.image, 0, imageSize);
            }
        } catch (IOException e) {
            return Packet.error(OPI_OPEN_FAILED + e);
        }

        if (parentScene != null) 
            Platform.runLater(()-> {
            textAreaCommands.appendText("OPI Initialized");
            textAreaCommands.appendText("\nPRL x " + this.prlx);
            textAreaCommands.appendText("\nPRL y " + this.prly);
            textAreaCommands.appendText("\nONH x " + this.onhx);
            textAreaCommands.appendText("\nONH y " + this.onhy);
            textAreaCommands.appendText("\n");
            if (this.image != null && this.settings.tracking) {
                Image image = new Image(new ByteArrayInputStream(this.image));
                imageView.setImage(image);
            }
            });

        return new Packet(queryResults());
    };
  
    /**
     * opiQuery: Query device
     * 
     * @return settings and state machine state
     *
     * @since 3.0.0
     */
    public Packet query() {
      String results = queryResults();

      if (parentScene != null)
        Platform.runLater(()-> {
          textAreaCommands.appendText("\nOPI Query:\n");
          textAreaCommands.appendText(results);
        });
      return new Packet(results);
    };
  
    /**
     * opiSetup: Change device background and overall settings
     * 
     * @param args pairs of argument name and value
     * 
     * @return A JSON object with return messages
     *
     * @since 3.0.0
     */
    @Parameter(name = "fixShape", className = Fixation.class, desc = "Fixation target type for eye.", optional = true, defaultValue = "SPOT")
    @Parameter(name = "fixCx", className = Double.class, desc = "x-coordinate of fixation target (degrees): Only valid values are -20, -6, -3, 0, 3, 6, 20 for fixation type 'spot' and -3, 0, 3 for fixation type 'square'.", optional = true, min = -20, max = 20, defaultValue = "0")
    @Parameter(name = "tracking", className = Double.class, desc = "Whether to correct stimulus location based on eye position.", optional = true, min = 0, max = 1, defaultValue = "0")
    public Packet setup(HashMap<String, Object> args) {
        if (!this.socket.isConnected()) return Packet.error(DISCONNECTED_FROM_HOST);

            // First the fixation marker
        int fixCx = (int) ((double) args.get("fixCx"));

        int fixShape = -1;
        switch((Fixation)args.get("fixShape")) {
            case SPOT -> {
              if (fixCx != 0 && Math.abs(fixCx) != 3 && Math.abs(fixCx) != 6 && Math.abs(fixCx) != 20)
                return Packet.error(String.format(INVALID_FIXATION_SETTING, fixCx, Fixation.SPOT));
              fixShape = 0;
            }
            case SQUARE -> {
              if (fixCx != 0 && Math.abs(fixCx) != 3)
                return Packet.error(String.format(INVALID_FIXATION_SETTING, fixCx, Fixation.SQUARE));
              fixShape = 1;
            }
        };
        try {
            this.send(OPI_SET_FIXATION + fixCx + " 0 " + fixShape);
            String result = this.readline();
            if (!result.equals("0"))
                return Packet.error(OPI_SET_FIXATION_FAILED + result);
        } catch (IOException e) {
            return Packet.error(OPI_SET_FIXATION_FAILED, e);
        }
        if (parentScene != null) 
        Platform.runLater(()-> {
            textAreaCommands.appendText("OPI Setup\n");
            textAreaCommands.appendText("\tSet Fixation to " + args.get("fixShape") + " at " + fixCx + "\n");
        });

            // And then the tracking
        int tracking = (int) ((double) args.get("tracking")); /// TODO might want to alter the canvas
        if (tracking != 0 && tracking != 1) 
            return Packet.error(INVALID_TRACKING_SETTING + tracking);
        try {
            this.send(OPI_SET_TRACKING + tracking);
            String result = this.readline();
            if (!result.equals("0"))
                return Packet.error(OPI_SET_TRACKING_FAILED + result);
            this.settings.tracking = true;
        } catch (IOException e) {
            return Packet.error(OPI_SET_TRACKING_FAILED, e);
        }
        if (parentScene != null) 
        Platform.runLater(()-> {
            textAreaCommands.appendText("\tTracking " + (this.settings.tracking ? "On" : "Off") + "\n");
        });

        return new Packet(queryResults());
    }
  
    /**
     * opiPresent: Present OPI stimulus in perimeter
     * 
     * @param args pairs of argument name and value
     * 
     * @return A JSON object with return messages
     *
     * @since 3.0.0
     */
    @Parameter(name = "x", className = Double.class, desc = "x co-ordinates of stimulus (degrees).", min = -30, max = 30, defaultValue = "0")
    @Parameter(name = "y", className = Double.class, desc = "y co-ordinates of stimulus (degrees).", min = -30, max = 30, defaultValue = "0")
    @Parameter(name = "lum", className = Double.class, desc = "Stimuli luminance (cd/m^2).", min = 0, max = 3183.099, defaultValue = "100")
    @Parameter(name = "t", className = Double.class, desc = "Presentation time (ms).", min = 200, max = 200, defaultValue = "200")
    @Parameter(name = "w", className = Double.class, desc = "Response window (ms).", min = 200, max = 2680, defaultValue = "1500")
    @ReturnMsg(name = "res", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
    @ReturnMsg(name = "res.eyex", className = Double.class, desc = "x co-ordinates of pupil at times eyet (pixels).")
    @ReturnMsg(name = "res.eyey", className = Double.class, desc = "y co-ordinates of pupil at times eyet (pixels).")
    @ReturnMsg(name = "res.eyed", className = Double.class, desc = "Diameter of pupil at times eyet (mm).")
    @ReturnMsg(name = "res.eyet", className = Double.class, desc = "Time of (eyex, eyey) pupil from stimulus onset (ms).", min = 0)
    @ReturnMsg(name = "res.time_rec", className = Double.class, desc = "Time since 'epoch' when command was received at Compass or Maia (ms).", min = 0)
    @ReturnMsg(name = "res.time_resp", className = Double.class, desc = "Time since 'epoch' when stimulus response is received, or response window expired (ms).", min = 0)
    @ReturnMsg(name = "res.num_track_events", className = Double.class, desc = "Number of tracking events that occurred during presentation.", min = 0)
    @ReturnMsg(name = "res.num_motor_fails", className = Double.class, desc = "Number of times motor could not follow fixation movement during presentation.", min = 0)
    public Packet present(HashMap<String, Object> args) {
        if (!this.socket.isConnected()) return Packet.error(DISCONNECTED_FROM_HOST);
        try {
            int level = (int) Math.round(-10 * Math.log10((double) args.get("lum") / (10000 / Math.PI)));
            StringBuilder opiMessage = new StringBuilder(OPI_PRESENT_STATIC).append(" ")
                .append((int) ((double) args.get("x"))).append(" ")
                .append((int) ((double) args.get("y"))).append(" ")
                .append(level).append(" 3 ")
                .append((int) ((double) args.get("t"))).append(" ")
                .append((int) ((double) args.get("w")));
                try {
                      this.send(opiMessage.toString());
                      updateGUIOnPresent(args);
                      return new Packet(parseResult(this.readline())); 
                } catch (IOException e) {
                      return Packet.error(OPI_PRESENT_FAILED, e);
                }
          } catch (ClassCastException | IllegalArgumentException e) {
          return Packet.error(OPI_PRESENT_FAILED, e);
        }
    }
  
    /**
     * opiClose: Close OPI connection
     * 
     * @param args pairs of argument name and value
     *
     * @return A JSON object with return messages
     *
     * @since 3.0.0
     */
    @ReturnMsg(name = "res.time", desc = "The time stamp for fixation data")
    @ReturnMsg(name = "res.x", desc = "The time stamp for fixation data")
    @ReturnMsg(name = "res.y", desc = "The time stamp for fixation data")
    public Packet close() {
        try {
            this.send(OPI_CLOSE);
           
            int n = this.incoming.readInt();
            if (n == 0 || n % 12 != 0) 
                return Packet.error("OPI_CLOSE did not receive the bytes it was expecting." );
            
            StringBuffer sb = new StringBuffer((3 + 6 + 5 + 5) * n / 12);   // guestimate of length
            for (int i = 0 ; i < n ; i += 12) {
                int t = this.incoming.readInt();
                float x = this.incoming.readFloat();
                float y = this.incoming.readFloat();
           
                sb.append("\n,{");
                sb.append(String.format("\"time\" : %s,", t));
                sb.append(String.format("\"posx\" : %s,", x));
                sb.append(String.format("\"posy\" : %s", y));
                sb.append("}");
            }
            sb.delete(0, 2); // remove leading "\n,"
           
            this.closeSocket();

            if (parentScene != null) 
            Platform.runLater(()-> {
                textAreaCommands.appendText("\nOPI Closed\n");
                textAreaCommands.appendText(sb.toString());
            });
            return new Packet(true, "[" + sb.toString() + "]");
        } catch (ClassCastException | IllegalArgumentException | IOException e) {
            return Packet.error(COULD_NOT_DISCONNECT, e);
        }
    };
  
    /**
     * Parse query results
     * 
     * @return A JSON object with query results
     *
     * @since 3.0.0
     */
    private String queryResults() {
        String image = "";
        if (this.image != null) {
            Base64.Encoder encoder = Base64.getEncoder();
            image = encoder.encodeToString(this.image);
        }
        return new StringBuilder("\n  {\n")
            .append("    \"prlx\": " + prlx + ",\n")
            .append("    \"prly\": " + prly + ",\n")
            .append("    \"onhx\": " + onhx + ",\n")
            .append("    \"onhy\": " + onhy + ",\n")
            .append("    \"image\": " + image + ",\n")
            .append("    \"minX\": " + settings.minX + ",\n")
            .append("    \"maxX\": " + settings.maxX + ",\n")
            .append("    \"minY\": " + settings.minY + ",\n")
            .append("    \"maxY\": " + settings.maxY + ",\n")
            .append("    \"minPressentationTime\": " + settings.minPressentationTime + ",\n")
            .append("    \"maxPressentationTime\": " + settings.maxPressentationTime + ",\n")
            .append("    \"minResponseWindow\": " + settings.minResponseWindow + ",\n")
            .append("    \"maxResponseWindow\": " + settings.maxResponseWindow+ ",\n")
            .append("    \"backgroundLuminance\": " + settings.backgroundLuminance + ",\n")
            .append("    \"minLuminance\": " + settings.minLuminance + ",\n")
            .append("    \"maxLuminance\": " + settings.maxLuminance + ",\n")
            .append("    \"tracking\": " + settings.tracking + "\n")
            .append("\n  }").toString();
    };
  
    /**
     * Parse results obtained for OPI-PRESENT-STATIC
     * 
     * @param received Message received from Icare machine
     * 
     * @return A JSON object with present results
     *
     * @since 3.0.0
     */
    private String parseResult(String received) {
      String[] message = received.split(" ");
      if (message[0] != "0") Packet.error(OPI_PRESENT_FAILED + "Error code received is: " + message[0]);
      return new StringBuilder("\n  {\n")
        .append("    \"seen\": " + message[1] + ",\n")
        .append("    \"time\": " + message[2] + ",\n")
        .append("    \"eyex\": " + message[9] + ",\n")
        .append("    \"eyey\": " + message[10] + ",\n")
        .append("    \"eyed\": " + message[8] + ",\n")
        .append("    \"eyet\": " + message[3] + ",\n")
        .append("    \"time_rec\": " + message[4] + ",\n")
        .append("    \"time_resp\": " + message[5] + ",\n")
        .append("    \"num_track_events\": " + message[6] + ",\n")
        .append("    \"num_motor_fails\": " + message[7] + ",\n  }").toString();
    }
}
